x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/chart_switch.tsx

```diff
@@ -189,28 +208,97 @@ export const ChartSwitch = memo(function ChartSwitch(props: Props) {
 
   const [searchTerm, setSearchTerm] = useState('');
 
-  const visualizationTypes = useMemo(
-    () =>
-      flyoutOpen &&
-      flatten(
-        Object.values(props.visualizationMap).map((v) =>
-          v.visualizationTypes.map((t) => ({
-            visualizationId: v.id,
-            ...t,
-            icon: t.icon,
-          }))
-        )
-      )
-        .filter(
-          (visualizationType) =>
-            visualizationType.label.toLowerCase().includes(searchTerm.toLowerCase()) ||
-            (visualizationType.fullLabel &&
-              visualizationType.fullLabel.toLowerCase().includes(searchTerm.toLowerCase()))
-        )
-        .map((visualizationType) => ({
-          ...visualizationType,
-          selection: getSelection(visualizationType.visualizationId, visualizationType.id),
-        })),
+  const { visualizationTypes, visualizationsLookup } = useMemo(
+    () => {
+      if (!flyoutOpen) {
+        return { visualizationTypes: [], visualizationsLookup: {} };
+      }
+      const activeVisualization = props.visualizationMap[props.visualizationId || ''];
+      const subVisualizationId = activeVisualization.getVisualizationTypeId(
+        props.visualizationState
+      );
+      const lowercasedSearchTerm = searchTerm.toLowerCase();
+      // reorganize visualizations in groups
+      const grouped: Record<
+        string,
+        Array<
+          VisualizationType & {
+            visualizationId: string;
+            selection: VisualizationSelection;
+          }
+        >
+      > = {};
+      // Will need it later on to quickly pick up the metadata from it
+      const lookup: Record<
+        string,
+        VisualizationType & {
+          visualizationId: string;
+          selection: VisualizationSelection;
+        }
+      > = {};
+      Object.entries(props.visualizationMap).forEach(([visualizationId, v]) => {
+        for (const visualizationType of v.visualizationTypes) {
+          const isSearchMatch =
+            visualizationType.label.toLowerCase().includes(lowercasedSearchTerm) ||
+            visualizationType.fullLabel?.toLowerCase().includes(lowercasedSearchTerm);
```

Comment:
So we're only looking at the i18n text for the search? I guess this makes sense...